core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -522,10 +522,11 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
         }
     }
 
-    public Engine.Index prepareIndexOnReplica(SourceToParse source, long seqNo, long version, VersionType versionType, long autoGeneratedIdTimestamp,
-                                              boolean isRetry) {
+    public Engine.Index prepareIndexOnReplica(SourceToParse source, long seqNo, long primaryTerm, long version, VersionType versionType,
+                                              long autoGeneratedIdTimestamp, boolean isRetry) {
         try {
             verifyReplicationTarget();
+            assert primaryTerm == this.primaryTerm : "op term [ " + primaryTerm + " ] != shard term [" + this.primaryTerm + "]";
```

Comment:
should this be a hard exception? something is bloody wrong here if this happens? maybe IllegalStateException?